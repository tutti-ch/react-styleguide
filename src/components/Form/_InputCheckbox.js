import React, { Component } from "react"
import PropTypes from "prop-types"
import { filterProps } from "helpers/functions"
import classes from "./Form.scss"
import WithWrapper from "./_WithWrapper"

export class InputCheckbox extends Component {
  static defaultProps = {
    checked: false
  }

  static propTypes = {
    /**
     * Whether the checkbox is checked or not.
     */
    checked: PropTypes.bool,

    /**
     * The value of the input.
     */
    value: PropTypes.any,

    /**
     * The label for the input.
     */
    label: PropTypes.string,
  }

  constructor(props) {
    super(props)

    this.state = {
      checked: props.checked,
    }

    this.toggle = this.toggle.bind(this)
  }

  /**
   * Sync the value.
   */
  componentWillReceiveProps(nextProps) {
    if (nextProps.checked !== this.props.checked) {
      this.setState({ checked: nextProps.checked })
    }
  }

  /**
   * Toggle the input checkbox.
   */
  toggle() {
    this.setState({ checked: !this.state.checked }, () => {
      if (typeof this.props.onChange === "function") {
        this.props.onChange(this.state.checked)
      }
    })
  }

  render() {
    const { value, checked } = this.state
    const { label } = this.props
    const props = filterProps(InputCheckbox.propTypes, this.props)

    return (
      // We do not use a real label here because react complains
      // about uncontrolled/controlled components.
      <span onClick={this.toggle} className={classes.cbLabel}>
        <input
          {...props}
          type="checkbox"
          value={value}
          checked={checked}
          onChange={this.toggle}
        />
        <span className={classes.labelText}>
          {label}
        </span>
      </span>
    )
  }
}

export default WithWrapper(InputCheckbox)
