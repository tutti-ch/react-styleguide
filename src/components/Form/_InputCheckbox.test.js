/* global describe, test, expect, jest */
import React from "react";
import { InputCheckbox } from "./_InputCheckbox";
import { shallow, mount } from "enzyme";

describe("(Component) InputCheckbox", () => {
  test("[componentWillReceiveProps] should sync the state with props", () => {
    const comp = shallow(<InputCheckbox />);
    const inst = comp.instance();
    inst.componentWillReceiveProps({ checked: true });
    expect(comp.state("checked")).toBe(true);
  });

  test("[toggle] should set the value and call the onChange from parent", () => {
    const spy = jest.fn();
    const comp = shallow(<InputCheckbox onChange={spy} />);
    const inst = comp.instance();
    inst.toggle();
    expect(comp.state("checked")).toBe(true);
    expect(spy).toHaveBeenCalled();
    inst.toggle.cancel();

    comp.setProps({ onChange: undefined });
    inst.toggle();
    expect(spy).toHaveBeenCalledTimes(1);
    expect(comp.state("checked")).toBe(false);
  });

  test("snapshot", () => {
    const comp = mount(
      <InputCheckbox value="Life is good" checked label="My label" />
    );
    expect(comp).toMatchSnapshot();
  });
});
