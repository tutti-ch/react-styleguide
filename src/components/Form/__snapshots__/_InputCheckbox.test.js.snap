// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Component) InputCheckbox snapshot 1`] = `
<InputCheckbox
  checked={true}
  label="My label"
  value="Life is good"
>
  <span
    onClick={[Function]}
  >
    <input
      checked={true}
      onChange={[Function]}
      type="checkbox"
    />
    <span>
      My label
    </span>
  </span>
</InputCheckbox>
`;
